# numbat.R

app <- "numbat.R"

args <- commandArgs(T)
if (length(args) < 9) {
  msg <- paste0("Usage: ", app, " <allele file> <count matrix dir>  \\
                                  <cell anno file> <ref cell type>  \\
                                  <output dir> <out prefix>   \\
                                  <genome version> <sequencing platform>  \\
                                  <ncores>")
  write(msg, file = stderr())
  quit("no", 1)
}

ale_fn <- args[1]
count_mtx_dir <- args[2]
cell_anno_fn <- args[3]
ref_cell_type <- args[4]
out_dir <- args[5]
out_prefix <- args[6]
genome <- args[7]
platform <- args[8]
ncores <- as.numeric(args[9])


library(Seurat)
library(numbat)

# check args
if (! dir.exists(out_dir)) {
  dir.create(out_dir, recursive = T)
}
setwd(out_dir)

if (! genome %in% c("hg19", "hg38"))
  stop(sprintf("invalid genome version '%s'.", genome))

gamma <- NULL
if (platform == "10x") {
  gamma <- 20
} else if (platform == "smartseq") {
  gamma <- 5
} else {
  stop(sprintf("invalid platform '%s'.", platform))
}

# filter ref cells from allele dataframe
ale <- read.delim(ale_fn, header = T, stringsAsFactors = F)
str(ale) 

cell_anno <- read.delim(cell_anno_fn, header = F, stringsAsFactors = F)
colnames(cell_anno) <- c("cell", "group")
ref_cells <- cell_anno[cell_anno$group %in% ref_cell_type, ]
str(ref_cells)

ale1 <- ale[! (ale$cell %in% ref_cells$cell), ]
str(ale1)
saveRDS(ale1, sprintf("%s/%s.ref_filtered.allele.dataframe.rds", out_dir, 
                          out_prefix))

# filter ref cells from count matrix
count_mtx <- Seurat::Read10X(data.dir = count_mtx_dir)   # gene x cell
str(count_mtx)

cnt_mtx1 <- count_mtx[, ! (colnames(count_mtx) %in% ref_cells$cell)]
str(cnt_mtx1)
saveRDS(cnt_mtx1, sprintf("%s/%s.ref_filtered.count.mtx.rds", out_dir,
                          out_prefix))

# calculate average expression of ref cells
ref_mean_expr <- numbat::aggregate_counts(count_mtx, ref_cells)
str(ref_mean_expr)
saveRDS(ref_mean_expr, sprintf("%s/%s.ref.gene_by_celltype.mtx.rds",
                               out_dir, out_prefix))

obj <- run_numbat(
  cnt_mtx1,          # gene x cell integer UMI count matrix 
  ref_mean_expr,     # reference expression profile, a gene x cell type normalized expression level matrix
  ale1,              # allele dataframe generated by pileup_and_phase script
  genome = genome,
  t = 1e-5,
  gamma = gamma,
  ncores = ncores,
  plot = TRUE,
  out_dir = out_dir
)

saveRDS(obj, sprintf("%s/%s.out.object.rds", out_dir, out_prefix))

print(paste0("[", app, "] All Done!"))