# Implementation
## afc
### Multi-gene UMIs
- Use pandas to load whole file in submodule `afc.mfu`, to speedup. 
  The divide-conquer strategy, which splits the cell or feature files into 
  small batches for parallel processing and then merge, can be time-consuming
  in large datasets, such as the HCC3 dataset.


### Read Assignment
- (Optional) assign a UMI to some gene only if certain fraction (e.g., 90%) 
  of the UMIâ€™s post-QC reads mapped to the gene. 
  It could be useful to help address the issue of multi-gene UMIs, 
  considering one UMI that 70% of its reads mapped to one gene, and the rest 
  30% reads mapped on a different gene.
    - add option `--minRAF` (minimum read assignment fraction)?
    - Other strategies can be used to process the above example 70%-30% UMI, 
      e.g., assign the UMI to the gene with more supporting reads and discard 
      the non-supporting reads, instead of using a hard cutoff of fraction.



## cs
- Write a wrapper function for cs module to make it an independent tool for
  count(RDR)-based CNA simulatioin, to take CNA profile and clone annotation
  files as input.

- Add an option about `total library size` to enable simulation of various
  read depth.
  Refer to the `total number of reads` in scDesign2.



# Discussion
### Should non-gene-assigned reads be included for read sampling?
In typical BAM files generated by CellRanger or SpaceRanger, reads can be
splitted into three groups:

- low-mapping-quality reads;

- (gene-assigned reads) high-mapping-quality reads that are assigned to 
  specific gene;
  
- (non-gene-assigned reads) high-mapping-quality reads that are not assigned
  to any genes, e.g., due to not compatible with gene structures;

The low-mapping-quality reads are always filtered before feature counting
or read sampling.
Current version (v0.5.0) only uses gene-assigned reads for feature counting
and read sampling, while excluding non-gene-assigned reads.
This strategy is similar as the one used in CellRanger and STARsolo counting, 
hence has the advantage that the UMI counts of `afc`, `cs`, and STARsolo 
counting on `rs` BAM, are close to each other.

However, including the non-gene-assigned reads can make the simulated BAM 
more like real BAM that contains some non-gene-assigned UMIs.
As the read sampling is random, then the simulated CNA signals should be 
unaffected after including those reads.

To include the non-gene-assigned UMIs,

- turn off the `xf` and `GN` tags by default, so that the Hap-A, B, and U
  UMIs contain both gene-assgined and non-gene-assigned UMIs.
  
- alternatively, extract and output non-gene-assigned UMIs for 3 (or 5) 
  haplotype states explicitly, in addition to the original gene-assigned
  UMIs of those states.



### Special UMIs
How to process some special UMIs in feature counting?

1. Within one UMI, reads overlap different features, e.g., due to errors in
   UMI collapsing, or features overlapping each other.

2. Within one UMI, some reads overlap one feature while others do not overlap
   any features.

3. Within one UMI, all reads overlap one feature, while some pass the
   --minINCLUDE filtering and others do not.



### Noisy signals in the reference cells in inferCNV heatmap.
Noise: both signals of copy gain and copy loss shown in a strip of genes.

1. process (e.g., remove) the special UMIs/reads listed above that the noise
   arises from.
   
2. process (e.g., remove) the special features (e.g., overlapping features) 
   that the noise arises from.
   
   

# Input



# Output



# Docs



# Tests
- Prepare notebooks & scripts to test each module.



# Scripts
